okhttp/src/main/java/okhttp3/ResponseBody.kt

```diff
@@ -0,0 +1,250 @@
+    }
+
+    val bytes: ByteArray = source().use(BufferedSource::readByteArray)
+    if (contentLength != -1L && contentLength != bytes.size.toLong()) {
+      throw IOException(
+          "Content-Length ($contentLength) and stream length (${bytes.size}) disagree")
+    }
+    return bytes
+  }
+
+  /**
+   * Returns the response as a character stream.
+   *
+   * If the response starts with a
+   * [Byte Order Mark (BOM)](https://en.wikipedia.org/wiki/Byte_order_mark), it is consumed and
+   * used to determine the charset of the response bytes.
+   *
+   * Otherwise if the response has a `Content-Type` header that specifies a charset, that is used
+   * to determine the charset of the response bytes.
+   *
+   * Otherwise the response bytes are decoded as `UTF-8`.
+   */
+  fun charStream(): Reader = reader
+
+  /**
+   * Returns the response as a string.
+   *
+   * If the response starts with a
+   * [Byte Order Mark (BOM)](https://en.wikipedia.org/wiki/Byte_order_mark), it is consumed and
+   * used to determine the charset of the response bytes.
+   *
+   * Otherwise if the response has a `Content-Type` header that specifies a charset, that is used
+   * to determine the charset of the response bytes.
+   *
+   * Otherwise the response bytes are decoded as `UTF-8`.
+   *
+   * This method loads entire response body into memory. If the response body is very large this
+   * may trigger an [OutOfMemoryError]. Prefer to stream the response body if this is a
+   * possibility for your response.
+   */
+  @Throws(IOException::class)
+  fun string(): String = source().use { source ->
+    source.readString(charset = Util.bomAwareCharset(source, charset()))
+  }
+
+  private fun charset() = contentType()?.charset(UTF_8) ?: UTF_8
+
+  override fun close() = Util.closeQuietly(source())
+
+  internal class BomAwareReader(
+    private val source: BufferedSource,
+    private val charset: Charset
+  ) : Reader() {
+
+    private var closed: Boolean = false
+    private var delegate: Reader? = null
+
+    @Throws(IOException::class)
+    override fun read(cbuf: CharArray, off: Int, len: Int): Int {
+      if (closed) throw IOException("Stream closed")
+
+      val finalDelegate = delegate ?: InputStreamReader(
+          source.inputStream(),
+          Util.bomAwareCharset(source, charset)).also {
+        delegate = it
+      }
+      return finalDelegate.read(cbuf, off, len)
+    }
+
+    @Throws(IOException::class)
+    override fun close() {
+      closed = true
+      delegate?.close() ?: run { source.close() }
+    }
+  }
+
+  companion object {
+
+    /**
+     * Returns a new response body that transmits [content]. If `contentType` is non-null
+     * and lacks a charset, this will use `UTF-8`.
+     */
+    @JvmStatic
+    fun create(contentType: MediaType?, content: String): ResponseBody {
+      val (charset, finalContentType) = contentType?.charset()
```

Comment:
This is too fancy for me. Could you just use regular control flow?

Reply:
Also – regular control flow doesn’t allocate!

Reply:
79622d0, though I kind of think this is fine aside from the extra allocation. The updated version feels less readable than the kotlin code since I'm now having to think about 2 levels of control flow, and default values on mutable possibly-rewritten variables